[AIEX] Scheduling heurististics for LCDs

gbossu opened the pull request:
This teaches the loop-aware scheduler a tiny bit about LCDs, and tries to reduce timing conflicts with the next iteration.

This gets us back to II=10 in Add2D, which had degraded to II=11 after #168 

```
| Core_Compute_Cycle_Count                                                  | Requantize_0 | Requantize_1 | Reciprocal_aie2_1 | Sign_int8_1  | Cast_aie2_int8 | Cast_aie2_int8_1 | EleMax_aie2_int8 | EleMin_aie2_int8 | Select_aie2_bf16 | Sign_int8_0  | HardSigmoidTemplated_bf16_0 | Shrink_aie2_0 |     | Conv1D_DW_AIE2_bf16_1 | BitwiseAnd_int8_0 | BitwiseOr_int8_0 | Conv1D_DW_AIE2_bf16_0 | AddAttributeBroadcasting_aie2_bf16 | SubAttributeBroadcasting_aie2_bf16_0 | AddBroadcastingBf16_aie2_0 | Add2D_Standalone_0 | Round_aie2_0 | SubBroadcasting_aie2_bf16_0 | AddBf16_aie2_0 | Sub_aie2_bf16_0 | Add2D_0      | Add2D_Standalone_1 | Abs_bf16_0   | Average diff |
| ------------------------------------------------------------------------- | ------------ | ------------ | ----------------- | ------------ | -------------- | ---------------- | ---------------- | ---------------- | ---------------- | ------------ | --------------------------- | ------------- |     | --------------------- | ----------------- | ---------------- | --------------------- | ---------------------------------- | ------------------------------------ | -------------------------- | ------------------ | ------------ | --------------------------- | -------------- | --------------- | ------------ | ------------------ | ------------ | ------------ |
| VITIS_AIE_QOR_MLLIB_loop_heuristics/20241004_153443_qor_report_peano.json | 1281         | 705          | 2400              | 122          | 761            | 761              | 209              | 209              | 328              | 416          | 616                         | 671           | ... | 4026                  | 482               | 482              | 3467                  | 787                                | 787                                  | 754                        | 334                | 381          | 732                         | 700            | 678             | 229          | 510                | 407          |              |
| VITIS_AIE_QOR_MLLIB_loop_heuristics/20241004_163728_qor_report_peano.json | 1343         | 735          | 2432              | 123          | 765            | 765              | 210              | 210              | 329              | 417          | 617                         | 672           |     | 3902                  | 467               | 467              | 3358                  | 760                                | 760                                  | 727                        | 322                | 367          | 705                         | 673            | 651             | 217          | 482                | 376          | -0.13%       |
| Total diff                                                                | REGR(+4.84%) | REGR(+4.26%) | REGR(+1.33%)      | REGR(+0.82%) | REGR(+0.53%)   | REGR(+0.53%)     | REGR(+0.48%)     | REGR(+0.48%)     | REGR(+0.30%)     | REGR(+0.24%) | REGR(+0.16%)                | REGR(+0.15%)  |     | IMPR(-3.08%)          | IMPR(-3.11%)      | IMPR(-3.11%)     | IMPR(-3.14%)          | IMPR(-3.43%)                       | IMPR(-3.43%)                         | IMPR(-3.58%)               | IMPR(-3.59%)       | IMPR(-3.67%) | IMPR(-3.69%)                | IMPR(-3.86%)   | IMPR(-3.98%)    | IMPR(-5.24%) | IMPR(-5.49%)       | IMPR(-7.62%) | -0.13%       |
```

gbossu committed:
[AIEX] NFC: Initialize interblock state when entering sched region

gbossu committed:
[AIEX] Basic heuristics for scheduling loops with LCDs

martien-de-jong commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -603,6 +595,13 @@ void InterBlockScheduling::enterRegion(MachineBasicBlock *BB,
+      // Now we can create the interblock edges between the top and the bottom
+      // region
+      BS.initInterBlock(*Context, *HR);
+    }
```
I don't think this makes sense for a multi-region loop. We will call the interblock init for each region in the block.

martien-de-jong replied:
I think everything would become clear if the first iteration of the fixpoint loop would only register the regions. Since we have a virtual schedule() now that shouldn't be too difficult to engineer.

gbossu replied:
Hmm, you're completely right. We indeed need a way to gather all regions before starting the actual scheduling. I'll work on that, maybe a new `GatheringRegions` stage in our sate machine.

martien-de-jong replied:
sate machine. Sounds delicious.

martien-de-jong commented on llvm/lib/Target/AIE/AIEMachineScheduler.cpp:
```diff
@@ -739,6 +766,29 @@ bool AIEPostRASchedStrategy::tryCandidate(SchedCandidate &Cand,
+    // loop-aware scheduler knows how to reassign those.
+    const BlockState &BS = getInterBlock().getBlockState(CurMBB);
+    if (UseLoopHeuristics && BS.Kind == AIE::BlockType::Loop &&
+        BS.getRegions().size() == 1 && BS.FixPoint.NumIters > 0) {
```
Yeah right. Note that getRegions().size() is unstable as long as BS.FixPoint.NumIters == 0

martien-de-jong commented on llvm/lib/Target/AIE/AIEMachineScheduler.cpp:
```diff
@@ -694,6 +698,29 @@ void AIEPostRASchedStrategy::handleRegionConflicts(
+/// this will be MAX_INT.
+int getEarliestLoopCarriedUse(const SUnit &SU,
+                              const InterBlockEdges &LoopEdges) {
+  const SUnit *SUInCurrIter = LoopEdges.getPreBoundaryNode(SU.getInstr());
```
I confuse `Iter` with Iterator. `SUInCurrentIteration`?

martien-de-jong commented on llvm/lib/Target/AIE/AIEMachineScheduler.cpp:
```diff
@@ -739,6 +766,29 @@ bool AIEPostRASchedStrategy::tryCandidate(SchedCandidate &Cand,
+                      Cand.SU->BotReadyCycle,
+                  TryCand, Cand, BotPathReduce)) {
+        return TryCand.Reason != NoCand;
+      }
```
AFAICS, this uses scheduling from the previous fixpoint iteration, which seems to me the best rationale not to trust the first scheduling run.

gbossu replied:
We do not really use the scheduling from the first iteration, only the "boundary DDG" and the height of nodes in the "post-boundary" half.

martien-de-jong replied:
Still.  BotReadyCycle sounds like a scheduling artifact to me. Or is this initialized with its height?

gbossu replied:
Oh sorry I mis-interpreted your comment. `getEarliestLoopCarriedUse(*TryCand.SU, LoopEdges)` is constant, but `Cand.SU->BotReadyCycle` obviously is not and will change as the loop schedule is converging. But if we want a number to fully trust, there's only the last iteration that will give it.

martien-de-jong commented on llvm/lib/Target/AIE/AIEMachineScheduler.cpp:
```diff
@@ -694,6 +698,29 @@ void AIEPostRASchedStrategy::handleRegionConflicts(
+    if (!LoopEdges.isPostBoundaryNode(Succ.getSUnit()))
+      continue;
+
+    EarliestCycle = std::min(EarliestCycle, int(Succ.getSUnit()->getHeight()));
```
Yes. BotReadyCycle is in Height currency, TopReadyCycle is in Depth currency. Let's introduce strong integer types for them.

andcarminati commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -856,6 +855,15 @@ const SUnit *InterBlockEdges::getPreBoundaryNode(MachineInstr *MI) const {
   return &DDG.SUnits.at(Found->second);
 }
 
+const SUnit *InterBlockEdges::getPostBoundaryNode(MachineInstr *MI) const {
```
I think this function is not used. Or do we have future plans for it?

gbossu replied:
Good catch, I ended up not needing it. Will remove

gbossu committed:
[AIEX] NFC: Ensure correct schedule() method is called for tiny regions

This isn't a problem yet, as ScheduleDAGMI::schedule() does more than
AIEScheduleDAGMI::schedule(), but who knows what the future holds.

gbossu commented:
Addressed the comments :) I also added one more commit because I realized we do not call `AIEScheduleDAGMI::schedule()` for regions of a single instruction. This is harmless for now, but I'd rather be cautious.

andcarminati reviewed (approved):
LGTM. Clean implementation that also makes one existing implicit state explicit.

martien-de-jong commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -928,6 +932,7 @@ void BlockState::classify() {
   if (LoopAware && IsLoop(TheBlock) &&
       llvm::all_of(TheBlock->successors(), CanFixLoopSchedule)) {
     Kind = BlockType::Loop;
+    FixPoint.Stage = SchedulingStage::GatheringRegions;
```
nit: I have a feeling that it would be more consistent to always gather regions in the first sweep, also for regular blocks But I agree that would introduce iterating on non-loop blocks. Not now.
